File: sdks/python/trace_harness/view/perspective.py
"""Display-tree projections for alternate trace perspectives."""

from __future__ import annotations

from dataclasses import dataclass

from trace_harness.model.node import Node
from trace_harness.view.display import DisplayNode
from trace_harness.view.facet import TracePerspective
from trace_harness.view.registry import FacetRegistry


@dataclass
class _ProjectedNode:
    display: DisplayNode
    hidden: int


def _source_count(display: DisplayNode) -> int:
    return max(len(display.node_ids), 1)


def _connector(hidden: int, node_ids: list[str], children: list[DisplayNode]) -> DisplayNode:
    return DisplayNode(
        kind="",
        name=f"… +{hidden} 个上下文节点",
        node_ids=list(dict.fromkeys(node_ids)),
        children=children,
    )


def project_perspective(
    roots: list[DisplayNode],
    by_id: dict[str, Node],
    registry: FacetRegistry,
    perspective: TracePerspective,
) -> list[DisplayNode]:
    """投影显示树；底层 Node.parent_node_id 始终只读，不生成第二套分析拓扑。"""
    if perspective == "full":
        return roots

    def visit(display: DisplayNode) -> _ProjectedNode | None:
        projected_children = [
            projected for child in display.children if (projected := visit(child))
        ]
        node = by_id.get(display.node_ids[0]) if display.kind and display.node_ids else None
        level = registry.perspective_level(node, perspective) if node else "detail"
        if level == "primary" or (level == "context" and projected_children):
            return _ProjectedNode(
                DisplayNode(
                    kind=display.kind,
                    name=display.name,
                    brief=display.brief,
                    node_ids=display.node_ids,
                    children=[projected.display for projected in projected_children],
                    findings=display.findings,
                    folded=0,
                ),
                hidden=0,
            )
        if not projected_children:
            return None

        own_hidden = _source_count(display)
        if len(projected_children) == 1 and projected_children[0].hidden > 0:
            child = projected_children[0]
            hidden = own_hidden + child.hidden
            return _ProjectedNode(
                _connector(
                    hidden,
                    [*display.node_ids, *child.display.node_ids],
                    child.display.children,
                ),
                hidden=hidden,
            )
        return _ProjectedNode(
            _connector(
                own_hidden,
                display.node_ids,
                [projected.display for projected in projected_children],
            ),
            hidden=own_hidden,
        )

    return [projected.display for root in roots if (projected := visit(root))]

File: sdks/typescript/trace-harness/src/view/perspective.ts
import type { Node } from "../model/node";
import type { DisplayNode } from "./display";
import type { TracePerspective } from "./facet";
import type { FacetRegistry } from "./registry";

interface ProjectedNode {
  display: DisplayNode;
  hidden: number;
}

function sourceCount(display: DisplayNode): number {
  return Math.max(display.node_ids.length, 1);
}

function connector(
  hidden: number,
  nodeIds: string[],
  children: DisplayNode[],
): DisplayNode {
  return {
    kind: "",
    name: `… +${hidden} 个上下文节点`,
    brief: [],
    node_ids: [...new Set(nodeIds)],
    children,
    findings: [],
    folded: 0,
  };
}

/**
 * Projects one immutable analysis tree into a presentation tree.
 *
 * Non-primary paths stay represented by synthetic connector rows, so the projection never
 * rewrites Node.parent_node_id or invents an alternative analysis topology.
 */
export function projectPerspective(
  roots: DisplayNode[],
  byId: Map<string, Node>,
  registry: FacetRegistry,
  perspective: TracePerspective,
): DisplayNode[] {
  if (perspective === "full") return roots;

  const visit = (display: DisplayNode): ProjectedNode | undefined => {
    const projectedChildren = display.children
      .map(visit)
      .filter((item): item is ProjectedNode => Boolean(item));
    const node = display.kind && display.node_ids.length
      ? byId.get(display.node_ids[0]!)
      : undefined;
    const level = node ? registry.perspectiveLevel(node, perspective) : "detail";
    if (level === "primary" || (level === "context" && projectedChildren.length)) {
      return {
        display: {
          ...display,
          children: projectedChildren.map((item) => item.display),
          folded: 0,
        },
        hidden: 0,
      };
    }
    if (!projectedChildren.length) return undefined;

    const ownHidden = sourceCount(display);
    if (projectedChildren.length === 1 && projectedChildren[0]!.hidden > 0) {
      const child = projectedChildren[0]!;
      const hidden = ownHidden + child.hidden;
      return {
        display: connector(
          hidden,
          [...display.node_ids, ...child.display.node_ids],
          child.display.children,
        ),
        hidden,
      };
    }
    return {
      display: connector(
        ownHidden,
        display.node_ids,
        projectedChildren.map((item) => item.display),
      ),
      hidden: ownHidden,
    };
  };

  return roots.map(visit)
    .filter((item): item is ProjectedNode => Boolean(item))
    .map((item) => item.display);
}
